File: projects/tango-contracts/smart_contracts/tango/contract.py
from algopy import ARC4Contract, String, UInt64, Txn, Account, BoxMap
from algopy.arc4 import abimethod, Struct


class Game(Struct):
    player1: Account
    player2: Account
    move1: UInt64
    move2: UInt64
    status: UInt64
    winner: Account


class Tango(ARC4Contract):
    """
    Tango - Rock Paper Scissors PvP Game
    A decentralized Rock Paper Scissors game on Algorand
    
    Move values: 0=None, 1=Rock, 2=Paper, 3=Scissors
    Status values: 0=Waiting, 1=Ongoing, 2=Finished
    """
    
    def __init__(self) -> None:
        self.owner = Txn.sender
        self.game_counter = UInt64(0)
        self.games = BoxMap(UInt64, Game)
    
    @abimethod()
    def create_game(self) -> UInt64:
        """Create a new game and return the game ID"""
        self.game_counter += 1
        game_id = self.game_counter
        
        # Initialize new game
        new_game = Game(
            player1=Txn.sender,
            player2=Account(),
            move1=UInt64(0),  # None
            move2=UInt64(0),  # None
            status=UInt64(0),  # Waiting
            winner=Account()
        )
        
        # Store game in box storage
        self.games[game_id] = new_game.copy()
        
        return game_id
    
    @abimethod()
    def join_game(self, game_id: UInt64) -> String:
        """Join an existing game as player 2"""
        assert game_id <= self.game_counter, "Invalid game ID"
        
        game = self.games[game_id].copy()
        
        # Validate game can be joined
        assert game.player1 != Account(), "Game does not exist"
        assert game.player2 == Account(), "Game is full"
        assert game.status == UInt64(0), "Game already started"
        
        # Add player 2 and update status
        game.player2 = Txn.sender
        game.status = UInt64(1)  # Ongoing
        
        # Update game in storage
        self.games[game_id] = game.copy()
        
        return String("Successfully joined game!")
    
    @abimethod()
    def submit_move(self, game_id: UInt64, move: UInt64) -> String:
        """Submit your move (1=Rock, 2=Paper, 3=Scissors)"""
        assert move >= 1 and move <= 3, "Invalid move (use 1=Rock, 2=Paper, 3=Scissors)"
        assert game_id <= self.game_counter, "Invalid game ID"
        
        game = self.games[game_id].copy()
        
        assert game.status == UInt64(1), "Game is not ongoing"
        
        # Determine which player is submitting
        if Txn.sender == game.player1:
            assert game.move1 == UInt64(0), "Player 1 already submitted move"
            game.move1 = move
        elif Txn.sender == game.player2:
            assert game.move2 == UInt64(0), "Player 2 already submitted move"
            game.move2 = move
        else:
            assert False, "You are not a player in this game"
        
        # Update game
        self.games[game_id] = game.copy()
        
        # Check if both moves are submitted
        if game.move1 != UInt64(0) and game.move2 != UInt64(0):
            return self._determine_winner(game_id)
        
        return String("Move submitted! Waiting for opponent...")
    
    @abimethod(readonly=True)
    def get_game(self, game_id: UInt64) -> Game:
        """Get game details by ID"""
        assert game_id <= self.game_counter, "Invalid game ID"
        return self.games[game_id].copy()
    
    @abimethod(readonly=True)
    def get_game_status(self, game_id: UInt64) -> String:
        """Get human-readable game status"""
        assert game_id <= self.game_counter, "Invalid game ID"
        game = self.games[game_id].copy()
        
        if game.status == UInt64(0):
            return String("Waiting for player 2")
        elif game.status == UInt64(1):
            return String("Game ongoing")
        else:
            if game.winner == Account():
                return String("Game finished - Draw!")
            else:
                return String("Game finished - Winner declared!")
    
    @abimethod()
    def _determine_winner(self, game_id: UInt64) -> String:
        """Internal method to determine the winner"""
        game = self.games[game_id].copy()
        
        # Check for draw
        if game.move1 == game.move2:
            game.winner = Account()
            game.status = UInt64(2)  # Finished
            self.games[game_id] = game.copy()
            return String("Game finished - It's a draw!")
        
        # Determine winner
        # Rock(1) beats Scissors(3)
        # Paper(2) beats Rock(1)
        # Scissors(3) beats Paper(2)
        player1_wins = (
            (game.move1 == UInt64(1) and game.move2 == UInt64(3)) or
            (game.move1 == UInt64(2) and game.move2 == UInt64(1)) or
            (game.move1 == UInt64(3) and game.move2 == UInt64(2))
        )
        
        if player1_wins:
            game.winner = game.player1
            result = String("Game finished - Player 1 wins!")
        else:
            game.winner = game.player2
            result = String("Game finished - Player 2 wins!")
        
        game.status = UInt64(2)  # Finished
        self.games[game_id] = game.copy()
        
        return result
    
    @abimethod(readonly=True)
    def get_winner(self, game_id: UInt64) -> Account:
        """Get the winner of a finished game"""
        assert game_id <= self.game_counter, "Invalid game ID"
        game = self.games[game_id].copy()
        assert game.status == UInt64(2), "Game not finished yet"
        return game.winner
    
    @abimethod()
    def transfer_ownership(self, new_owner: Account) -> String:
        """Transfer contract ownership"""
        assert Txn.sender == self.owner, "Only owner can transfer ownership"
        assert new_owner != Account(), "Invalid new owner address"
        
        self.owner = new_owner
        return String("Ownership transferred successfully!")

File: projects/tango-frontend/src/constants/contract.ts
// Contract configuration
// These values need to be updated after deploying your smart contract
export const CONTRACT_CONFIG = {
  // Replace with your actual deployed contract app ID
  appId: Number(import.meta.env.VITE_CONTRACT_APP_ID) || 748969226,

  // Replace with your actual contract app address (calculated after deployment)
  appAddress: import.meta.env.VITE_CONTRACT_APP_ADDRESS || 'DHGTKVJVMIJ3KZHUXQFJTJK3CZ5ULIWIV2VJIKFGIA44E4PKGVDM4JAZPY',

  // Contract ABI - Full ABI Contract object for method calls
  contract: {
    name: "Tango",
    desc: "Rock Paper Scissors PvP Game",
    methods: [
      {
        name: "create_game",
        desc: "Create a new game and return the game ID",
        args: [],
        returns: { type: "uint64", desc: "Game ID" }
      },
      {
        name: "join_game",
        desc: "Join an existing game as player 2",
        args: [{ type: "uint64", name: "game_id", desc: "Game ID to join" }],
        returns: { type: "string", desc: "Success message" }
      },
      {
        name: "submit_move",
        desc: "Submit your move (1=Rock, 2=Paper, 3=Scissors)",
        args: [
          { type: "uint64", name: "game_id", desc: "Game ID" },
          { type: "uint64", name: "move", desc: "Move: 1=Rock, 2=Paper, 3=Scissors" }
        ],
        returns: { type: "string", desc: "Game result or waiting message" }
      },
      {
        name: "get_game",
        desc: "Get game details by ID",
        args: [{ type: "uint64", name: "game_id", desc: "Game ID" }],
        returns: { type: "(address,address,uint64,uint64,uint64,address)", desc: "Game struct" }
      },
      {
        name: "get_game_status",
        desc: "Get human-readable game status",
        args: [{ type: "uint64", name: "game_id", desc: "Game ID" }],
        returns: { type: "string", desc: "Status message" }
      },
      {
        name: "_determine_winner",
        desc: "Internal method to determine the winner",
        args: [{ type: "uint64", name: "game_id", desc: "Game ID" }],
        returns: { type: "string", desc: "Winner message" }
      },
      {
        name: "get_winner",
        desc: "Get the winner of a finished game",
        args: [{ type: "uint64", name: "game_id", desc: "Game ID" }],
        returns: { type: "address", desc: "Winner address" }
      },
      {
        name: "transfer_ownership",
        desc: "Transfer contract ownership",
        args: [{ type: "address", name: "new_owner", desc: "New owner address" }],
        returns: { type: "string", desc: "Success message" }
      }
    ]
  },

  // Contract ABI methods (Rock-Paper-Scissors Game)
  methods: {
    createGame: 'create_game',
    joinGame: 'join_game',
    submitMove: 'submit_move',
    getGame: 'get_game',
    getGameStatus: 'get_game_status',
    determineWinner: '_determine_winner',
    getWinner: 'get_winner',
    transferOwnership: 'transfer_ownership'
  },

  // Box storage prefix for games
  boxPrefix: 'games',

  // Schema (for deployment tools)
  globalInts: 1,
  globalBytes: 1,
  localInts: 0,
  localBytes: 0
}

// Game status enum (for UI clarity)
export const GAME_STATUS = {
  WAITING: 0,
  ONGOING: 1,
  FINISHED: 2
} as const

export const GAME_STATUS_LABEL: Record<number, string> = {
  0: 'Waiting for Player 2',
  1: 'Game in Progress',
  2: 'Game Finished'
}

// Move options
export const MOVE = {
  NONE: 0,
  ROCK: 1,
  PAPER: 2,
  SCISSORS: 3
} as const

export const MOVE_LABEL: Record<number, string> = {
  0: 'None',
  1: 'Rock',
  2: 'Paper',
  3: 'Scissors'
}

// Network configuration
export const NETWORK_CONFIG = {
  TESTNET: {
    name: 'TestNet',
    algodServer: 'https://testnet-api.algonode.cloud',
    algodPort: '',
    algodToken: '',
    indexerServer: 'https://testnet-idx.algonode.cloud',
    indexerPort: '',
    indexerToken: ''
  },
  MAINNET: {
    name: 'MainNet',
    algodServer: 'https://mainnet-api.algonode.cloud',
    algodPort: '',
    algodToken: '',
    indexerServer: 'https://mainnet-idx.algonode.cloud',
    indexerPort: '',
    indexerToken: ''
  },
  LOCALNET: {
    name: 'LocalNet',
    algodServer: 'http://localhost',
    algodPort: '4001',
    algodToken: 'a'.repeat(64),
    indexerServer: 'http://localhost',
    indexerPort: '8980',
    indexerToken: ''
  }
}

// Transaction fees and limits (in microAlgos)
export const TRANSACTION_FEES = {
  MIN_FEE: 1000, // 0.001 ALGO
  APP_CALL_FEE: 1000,
  BOX_ACCESS_FEE: 400, // Per box read/write
  MIN_ACCOUNT_BALANCE: 100000, // 0.1 ALGO
  MIN_APP_BALANCE: 100000, // Base app balance
  MIN_BOX_BALANCE: 28500 // Per box (0.0285 ALGO)
}

// Error messages
export const ERROR_MESSAGES = {
  WALLET_NOT_CONNECTED: 'Please connect your wallet to continue',
  INSUFFICIENT_BALANCE: 'Insufficient balance to complete this transaction',
  TRANSACTION_FAILED: 'Transaction failed. Please try again.',
  NETWORK_ERROR: 'Network error. Please check your connection.',
  CONTRACT_NOT_FOUND: 'Contract not found. Please check the app ID.',
  NOT_AUTHORIZED: 'You are not authorized to perform this action',
  INVALID_GAME_ID: 'Invalid game ID',
  GAME_NOT_FOUND: 'Game not found or does not exist',
  GAME_FULL: 'This game already has two players',
  GAME_NOT_WAITING: 'Game is not waiting for a player',
  GAME_ALREADY_STARTED: 'Game has already started',
  GAME_FINISHED: 'Game is already finished',
  INVALID_MOVE: 'Invalid move. Choose Rock, Paper, or Scissors.',
  MOVE_ALREADY_SUBMITTED: 'You have already submitted your move',
  SAME_PLAYER: 'Cannot play against yourself',
  INVALID_ADDRESS: 'Invalid address format'
}

// Success messages
export const SUCCESS_MESSAGES = {
  GAME_CREATED: 'Game created! Share the Game ID with your opponent.',
  GAME_JOINED: 'Successfully joined the game!',
  MOVE_SUBMITTED: 'Move submitted! Waiting for opponent...',
  GAME_FINISHED: 'Game over! Check the result.',
  WALLET_CONNECTED: 'Wallet connected successfully!',
  TRANSACTION_SENT: 'Transaction sent successfully!',
  OWNERSHIP_TRANSFERRED: 'Ownership transferred successfully!'
}
